### PR TITLE
chore(files): Revamps Elgg\Filesystem with Flysystem

### LINKED ISSUE #8467: GaufretteDirectory::getFiles() is slow
No matter your Directory path, `getFiles()` must scan the entire filesystem because it uses `Gaufrette\Filesystem::listKeys`. E.g. if you make a local filesystem at Elgg's root, even if you `->chroot()` into subdirectories, getFiles() will scan the entire Elgg filesystem including `vendor` and `.git`.

### LINKED ISSUE #8468: Elgg\Filesystem\Directory needs a way to list only children
getFiles() automatically returns the whole tree. Maybe add a `$recursive = true` argument?

### PR DESCRIPTION
**Pull request by mrclay**
This adds Directory methods for finding files, directories, or both, and replaces the underlying implementation with Flysystem.

Fixes #8468
Fixes #8467

### CODE REVIEW

File: engine/classes/Elgg/Filesystem/Directory.php [hunk 1]
```diff
@@ -36,17 +42,46 @@ public function getContents($path);
...
+	 *
+	 * @param string $path      The subdirectory path within this directory
+	 * @param bool   $recursive Find files and directories recursively
+	 *
+	 * @return Collection<File|Directory>
+	 *
+	 * @throws \InvalidArgumentException
 	 */
-	public function getFiles($path = '');
+	public function getEntries($path = '', $recursive = true);
```
Comment (ewinslow) [inline engine/classes/Elgg/Filesystem/Directory.php]: I wouldn't think this should be part of the interface.

Comment (mrclay) [inline engine/classes/Elgg/Filesystem/Directory.php]: Why?

Comment (ewinslow) [inline engine/classes/Elgg/Filesystem/Directory.php]: There is no present use case for it.

**Comment by ewinslow** [comment]
Still had to change a lot of gaufrette references to fly system in tests and such... We should not expose those details to our clients if we can avoid it...

**Comment by ewinslow** [comment]
A pattern I have taken elsewhere but we're not super consistent with:

```
Elgg\{Namespace}\{Interface}\{Implementation}::from{ArgType}
```

For example, in this case we could have:

```
Elgg\Filesystem\Directory\InMemory::fromArray
Elgg\Filesystem\Directory\Local::fromPath
```

Instead of the current \{ThirdParty}{Interface}::create{Implementation}

And that way we we'd have even less leakage of the third party library backing our filesystem in our tests and such.

**Comment by ewinslow** [comment]
I see this same {Vendor}\{Namespace}\{Interface}\{Implementation} pattern frequently in other PHP libs too. Flysystem being one of them, in fact!

**Comment by ewinslow** [comment]
In any case, this is a good change and fixes 2 bugs. The only thing I'd block on is the new interface method `getEntries`. We should discuss separately whether adding that is necessary.

**Comment by mrclay** [comment]
Removed Directory::getEntries and made it protected.

**Comment by ewinslow** [comment]
LGTM

**Pull request merged**